### PR TITLE
Release/0.4.8 Refactor banking operations and transaction history with improved validation title

### LINKED ISSUE #247: add integration tests for `Banking Operations System`

### LINKED ISSUE #248: Refactor TransferBuilder: Remove Deprecated Transaction Type Methods
### **Description:**

The `TransferBuilder` class currently contains several methods marked as `@Deprecated` for setting the transaction type:

*   `withTransactionType(String type)`
*   `withTransactionType(TransactionType type)`
*   `asOwnTransfer()`
*   `asInternalTransfer()`
*   `asExternalTransfer()`

These methods are no longer necessary because the transaction type (`TRANSFER_OWN`, `TRANSFER_INTERNAL`, `TRANSFER_EXTERNAL`) is now determined internally within the `build()` method by the `resolveTransactionType()` logic, based on the source and destination accounts.

### **Task:**

1.  Identify all usages of the deprecated methods listed above throughout the codebase.
2.  Refactor the code to remove calls to these deprecated methods. The `TransferBuilder` should rely solely on its internal logic to set the correct transaction type.
3.  Remove the deprecated methods from the `TransferBuilder` class.
4.  Ensure all relevant tests still pass after the refactoring.

This cleanup will simplify the `TransferBuilder` API and remove potential confusion by eliminating outdated methods.

### PR DESCRIPTION
**Pull request by Pawel-Mackiewicz**
---

### Description of Changes

This pull request includes several updates aimed at improving the banking and transaction systems:

- **Documentation Updates**: Enhanced and refined system documentation for clearer architecture, API details, validation processes, exception handling, and integration points.
- **System Refactoring**:
  - Improvements in `TransferBuilder`, transaction type handling, and validation processes.
  - Enhanced modularity and maintainability by restructuring packages and API paths.
  - Removed deprecated classes and fields for a cleaner codebase.
  - Introduced generic abstractions and centralized security logic for better flexibility.
- **Testing Enhancements**: Added comprehensive integration tests covering edge cases, unauthorized access, service utilities, and API response validations.
- **Security and Maintenance**:
  - Improved DTOs with JSON-specific annotations.
  - Strengthened validations and error handling.
  - Enhanced API response correctness and documentation accuracy.

These changes aim to improve the overall system reliability, extensibility, and developer experience.

### Related Issue(s)

- Closes #247
- Closes #248

### Checklist

- [ ] Documentation has been updated as needed.
- [ ] Tests have been added or updated to cover the changes.
- [ ] Related issues have been linked in this PR.

## Summary by CodeRabbit

- **New Features**
  - Introduced abstract and concrete account authorization services for IBAN and ID-based ownership validation.
  - Added REST API interfaces and controllers for banking operations and transaction history with security and detailed OpenAPI documentation.
  - Added paginated and filterable transaction history endpoints with export capabilities in CSV and PDF formats.
  - Introduced transaction precondition validator to enforce sufficient funds before transfers.
  - Added integration test utilities for user and account creation to support comprehensive testing.

- **Improvements**
  - Refactored transaction and transfer services to remove explicit user ownership checks, improve DTO usage, and simplify method signatures.
  - Enhanced transaction filtering with status support and improved sorting options.
  - Updated frontend transaction history scripts to match new API endpoints and data structures.
  - Improved exception handling and validation logic, including email format validation before account existence checks.
  - Enhanced logging and error message clarity across services.

- **Bug Fixes**
  - Fixed validation and error response handling in email validation and transfer operations.
  - Corrected transaction type resolution and removed redundant or conflicting transaction type setters.

- **Documentation**
  - Added extensive documentation for API usage, banking operations, exception handling, notifications, registration, token system, transaction history, and transaction processing.

- **Tests**
  - Added new integration tests for banking operations and transaction history APIs covering success, error, and security scenarios.
  - Refactored existing tests to use dedicated integration test services and updated to reflect service signature changes.
  - Added unit tests for new authorization services.

- **Chores**
  - Updated project version to 0.4.8 and corresponding build artifacts.
  - Removed deprecated classes and interfaces related to transaction history and banking operations.
  - Reorganized packages and imports for better modularity and maintainability.

- **Refactor**
  - Moved and renamed DTOs, services, and controllers to align with new package structure emphasizing banking operations and history.
  - Simplified transaction builder by automating transaction type resolution.
  - Removed internal transfer markings from welcome bonus transactions.
  - Streamlined service method signatures by removing unnecessary parameters and dependencies.
  - Consolidated exception handling and logging strategies for clearer error management.

### CODE REVIEW

File: src/test/java/info/mackiewicz/bankapp/shared/service/IbanAnalysisServiceTest.java [hunk 1]
```diff
@@ -1,18 +1,15 @@
 package info.mackiewicz.bankapp.shared.service;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertFalse;
-import static org.junit.jupiter.api.Assertions.assertTrue;
-
+import info.mackiewicz.bankapp.account.util.IbanGenerator;
+import info.mackiewicz.bankapp.system.banking.operations.service.helpers.IbanAnalysisService;
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/test/java/info/mackiewicz/bankapp/shared/service/IbanAnalysisServiceTest.java:4]: done

File: src/main/java/info/mackiewicz/bankapp/system/banking/operations/api/dto/IbanTransferRequest.java [hunk 2]
```diff
@@ -1,29 +1,37 @@
...
 
+    @JsonIgnore
     public Iban getRecipientIban() {
         return Iban.valueOf(recipientIban);
     }
 
+    @JsonIgnore
     public void setRecipientIban(Iban iban) {
         this.recipientIban = iban.toString();
     }
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/banking/operations/api/dto/IbanTransferRequest.java:32]: @coderabbitai
deleted String setter. now json will use field for deserialization.

File: src/main/java/info/mackiewicz/bankapp/system/banking/history/controller/TransactionHistoryRestController.java [hunk 3]
```diff
@@ -1,52 +1,51 @@
...
+    @PreAuthorize("@idAccountAuthorizationService.validateAccountOwnership(#filter.accountId, authentication.principal)")
     @GetMapping
-    public ResponseEntity<Page<Transaction>> getTransactions(
-            @AuthenticationPrincipal User user, 
-            @ModelAttribute TransactionFilterDTO filter
+    @Override
+    public ResponseEntity<Page<TransactionResponse>> getTransactions(
+            @AuthenticationPrincipal User user,
+            @ModelAttribute @Valid TransactionFilterRequest filter
     ) {
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/banking/history/controller/TransactionHistoryRestController.java:34]: not true.

File: src/main/java/info/mackiewicz/bankapp/system/banking/operations/service/helpers/TransactionPreconditionValidator.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+     * @param sourceAccount The account from which funds are being withdrawn. Must not be null.
+     * @param amount        The amount to be checked against the account balance. Must not be null.
+     *
+     * @throws InsufficientFundsException If the account balance is less than the specified amount.
+     */
+    public void checkFunds(@NonNull Account sourceAccount, @NonNull BigDecimal amount) throws InsufficientFundsException {
+        if (sourceAccount.getBalance().compareTo(amount) < 0) {
+            throw new InsufficientFundsException("Insufficient funds");
+        }
+    }
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/banking/operations/service/helpers/TransactionPreconditionValidator.java:31]: it's done in dto.

File: src/main/java/info/mackiewicz/bankapp/system/banking/operations/service/TransferOperationService.java [hunk 5]
```diff
@@ -1,97 +1,99 @@
...
-        Transaction transfer = createTransferTransaction(request, validatedSourceAccount, destinationAccount);
+            preconditionValidator.checkFunds(sourceAccount, request.getAmount());
 
-        log.debug("Registering transfer transaction");
-        Transaction registeredTransaction = transactionService.registerTransaction(transfer);
+            log.debug("Fetching destination account");
+            Account destinationAccount = destinationAccountSupplier.get();
 
-        log.info("Transaction registered with ID: {}", registeredTransaction.getId());
+            log.debug("Creating transfer transaction");
```
[the remark this answers is not in the document]
Comment (Pawel-Mackiewicz) [reply src/main/java/info/mackiewicz/bankapp/system/banking/operations/service/TransferOperationService.java:64]: exception handling is done in `AccountQueryService` which is called in this method.

**Comment by Pawel-Mackiewicz** [comment]
merged from develop.
